---
title: "Convert Seurat to Python Object"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
I'm trying to use `phateR`, but keep on running into errors when transforming the data. So, I want to see if the Python package for PHATE is more developed than the R package. To do that, I need to convert the Seurat object to .h5ad.

```{r}
rm(list=ls())
library(Seurat)
library(ggplot2)
library(scater)
library(tidyverse)

#install.packages("devtools")
#devtools::install_github(repo="hhoeflin/hdf5r")
#devtools::install_github(repo="mojaveazure/loomR", 
#                         ref="develop")
#library(loomR)
```

# 1. Save Entire Data As loom File
```{r}
rm(list=ls())

# Load Seurat object as is
load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)
lung.loom = Seurat::as.loom(lung,filename="C:/Users/scott/Data/scRNASeq/shea/lung.loom")
```

# 2. Save Entire Dataset as .csv Files
Now we'll save the count data, the gene symbols, and the cell type as several csv files.
```{r}
rm(list=ls())

#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)
load("D:/Data/scRNASeq/shea/lung.robj", verbose=TRUE)


# Count data
data = lung@assays$RNA@counts
data = as.data.frame(data)
write_delim(data, 
            "D:/Data/scRNASeq/shea/lung.csv",
            delim=',')



# Gene Symbol Data
#MGI.Symbol = as.data.frame(rownames(data))
#write_delim(MGI.Symbol, 
#            "C:/Users/scott/Data/scRNASeq/shea/symbol.csv",
#            delim=",")

# Cell identification data
#active_indent = as.data.frame(lung@active.ident)
#active_indent$true_ident = rownames(active_indent)
#rownames(active_indent) = NULL
#write_delim(x=active_indent,
#            file="C:/Users/scott/Data/scRNASeq/shea/active_ident.csv",
#            delim=',')
```

# 3. Save Subsets as Separate Sheets in an Excel file
To perform data imputation and to save on memory, let's separate the data by cell types.
```{r}
rm(list=ls())

# ACLX
#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)
# Dell
load("D:/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

celltype.list = SplitObject(lung, split.by="CellType")
```

```{r}
# Make a function that maps genes using two reference databases
convertMouseGeneList = function(gene_obj, db1, db2){
  orthologs = getLDS(attributes=c("mgi_symbol"), filters="mgi_symbol", 
                   values=rownames(gene_obj), mart=db2, 
                   attributesL=c("hgnc_symbol", 
                                 "entrezgene_id", 
                                 "ensembl_gene_id"), martL=db1, 
                   uniqueRows=TRUE)
  mapped_gene_obj = merge(orthologs, gene_obj, 
                          by.x='MGI.symbol', by.y='row.names')
  return(mapped_gene_obj)
}
```

```{r}
install.packages('BiocManager')
BiocManager::install("biomaRt", 
                     dependencies=TRUE,
                     INSTALL_opts='--no-lock')
library(biomaRt)

# Create the mouse and human database
human = useMart("ensembl", dataset="hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset="mmusculus_gene_ensembl")
```

```{r}
data = list()
for(i in 1:length(celltype.list)){
  tmp = data.frame(celltype.list[[i]]@assays$RNA@counts)
  colnames(tmp) = celltype.list[[i]]@meta.data$EVERYTHING
  tmp = convertMouseGeneList(tmp, human, mouse)
  tmp$Ensembl = tmp$Gene.stable.ID
  data[[i]] = tmp
}
```

Save all.
```{r}
#install.packages('writexl')
writexl::write_xlsx(
  x=list(Stromal           = data[[1]],
       AlveolarMacrophages = data[[2]],
       Macrophages         = data[[3]],
       Monocytes           = data[[4]],
       Pneumocytes         = data[[5]],
       Bcells              = data[[6]],
       Granulocytes        = data[[7]],
       Tcells              = data[[8]],
       Dendritic           = data[[9]],
       Neutrophils         = data[[10]],
       NK                  = data[[11]],
       Endothelial         = data[[12]],
       HSC                 = data[[13]]),
  path="D:/Data/scRNASeq/shea/celldf_ensembl.xlsx")
#path="C:/Users/scott/Data/scRNASeq/shea/celldf.xlsx")

```