---
title: "Dimension reduction with Flux Balance Analysis"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
This notebook maps the metabolic fluxes from single-cell Flux Balance Analysis onto the dimension reduction embedding.

# 1. Read in libraries we'll use for downstream analyses
```{r}
library(tidyverse)
library(umap)
library(Seurat)
library(writexl)
```

# 2. Read in metabolic flux data and map to original cell ids
We'll read in the metabolic flux data computed from single-cell Flux Balance Analysis
```{r}
rm(list=ls())
flux_data = read_csv('D:/Analysis/Immune/neutrophil_flux.csv')
flux_data[, 2:ncol(flux_data)] = abs(flux_data[, 2:ncol(flux_data)])
```

Now let's perform some mapping to the original ids
```{r}
# Read in original data
ids = read_csv("D:/Data/scRNASeq/shea/neutrophil_colmap.csv")

# Grab day
ids$day = ids$final_col %>%
    strsplit(" ") %>%
    sapply(tail, 1)

# Iteratively add 1 for all duplicates
ids$day = as.numeric(ids$day)
ids$day = ids$day * 10
ids = data.frame(ids)
ids$new_day = ave(ids$day, ids$final_col, FUN = function(x) x[1] + seq_along(x) - 1)
ids$new_day = ids$new_day - ids$day
ids$day = ids$day / 10

# Create the new ID
ids$final_day = paste(ids$day, ids$new_day, sep='.')
ids$final_day = as.character(ids$final_day)

# Manually set the first call for a given day
ids$final_day[1] = '0'
ids$final_day[2] = '14'
ids$final_day[19] = '21'
ids$final_day[77] = '7'

# Create the final output
ids$final_col = paste("Neutrophils", ids$final_day, sep="  ")
```

We have to now merge the ids and the metabolic flux data to get a cohesive mapping of cell identifiers to flux.
```{r}
mapped_flux = merge(ids, 
                    flux_data, 
                    by.x='final_col', 
                    by.y='ID')
```

# Load embeddings
First, let's load the Seurat object.
```{r}
#rm(list=ls())
# Load the data
# Dell
load("D:/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# ACLX
#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# Linux
#load("~/Data/scRNASeq/shea/lung.Robj")
#unique(neutrophils@meta.data[["CellType"]])
```

We'll now subset the neutrophil data
```{r}
orig.indent = lung@active.ident
celltype.list = SplitObject(lung, split.by="CellType")

# Create separate Seurat objects to visualize.
neutrophils   = celltype.list$Neutrophils
```

This function will find subpopulations within a given cell type. This can be used later.
```{r}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  # Log normalize data
  gene_obj = NormalizeData(gene_obj, 
                       normalization.method='LogNormalize', 
                       scale.factor=10000)
  
  # Scale data
  all.genes = rownames(gene_obj)
  gene_obj = ScaleData(gene_obj, 
                   features=all.genes)
  
  gene_obj = SCTransform(gene_obj)

  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                npcs=50,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:50,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  gene_obj = RunTSNE(gene_obj,
                     dims=1:50)
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:50)
  
  return(gene_obj)
}
```

## T-SNE
Now let's visualize the original T-SNE plot. First, we'll cluster the data. 
```{r}
neutrophils = FindNeighbors(neutrophils, 
                       dims=1:50,
                       reduction='pca')
neutrophils = FindClusters(neutrophils, 
                    resolution=0.25)
```

Then we'll overlay the clusters onto the current tSNE embedding.
```{r}
# Plot based on predicted clusters
DimPlot(neutrophils, 
        reduction="tsne",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="tSME 1", y="tSNE 2") +
        theme_minimal() +
        theme(plot.title = element_text(size=25),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

Knowing the current cluster set, we should map the clusters to the flux data. Then we can compute differentially active metabolic flux profiles based on subclusters.

First, we save the tsne clusters.
```{r}
tsne_cluster_ident = as.factor(neutrophils@active.ident)
```

Then, we create the metabolic flux object.
```{r}
# Create identity
mapped_flux = mapped_flux[, -c(final_col, day, new_day, final_day)]
# Cells need to be cols and rows need to be reactions
flux_data = t(flux_data)

# Set column names
#names(flux_data) = lapply(flux_data[1, ], as.character)
colnames(flux_data) = as.character(flux_data["ID", ])
flux_data = flux_data[-1, ] 
```


## UMAP
```{r}
neutrophils = find_subpopulations(neutrophils)

# Plot based on predicted clusters
DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Unsupervised learning") +
        theme_minimal() +
        theme(plot.title = element_text(size=25),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

Now that we know everything is working, let's grab the data.

```{r}
umap_data = data.frame(neutrophils@reductions$umap@cell.embeddings)
umap_data$cluster = neutrophils@active.ident
merged_umap = merge(umap_data, mapped_flux, by.x=0, by.y='orig_col')
```

# Visualize flux data onto original UMAP embedding

```{r}
ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=cluster))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Unsupervised clusters")
```
```{r}
p1 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_3802))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Glutamate dehydrogenase") + 
        scale_color_gradient(low="gray", high="blue")

p2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_8486))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenosine Phosphorylase") + 
        scale_color_gradient(low="gray", high="blue")

p3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4012))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenosine Kinase") + 
        scale_color_gradient(low="gray", high="blue")

p4 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4085))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenine Phosphoribosyltransferase") + 
        scale_color_gradient(low="gray", high="blue")

scater::multiplot(p1, p2, p3, p4, cols=2)
```

```{r}
q1 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4418))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Hypoxanthine Phosphoribosyltransferase") + 
        scale_color_gradient(low="gray", high="red")

q2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4574))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Inosine Phosphorylase") + 
        scale_color_gradient(low="gray", high="red")

q3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=INSK))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Inosine Kinase") + 
        scale_color_gradient(low="gray", high="red")

q4 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4586))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Isocitrate:NADP+ Oxidoreductase") + 
        scale_color_gradient(low="gray", high="red")

q5 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4588))+ 
        geom_point() +
        theme_minimal() +
        labs(title="Isocitrate:NAD+ Oxidoreductase") + 
        scale_color_gradient(low="gray", high="red")
scater::multiplot(q1, q2, q3, q4, q5, cols=2)
```

```{r}
q1 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4418))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Hypoxanthine Phosphoribosyltransferase") + 
        scale_color_gradient(low="gray", high="red")

q2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4574))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Inosine Phosphorylase") + 
        scale_color_gradient(low="gray", high="red")

q3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=INSK))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Inosine Kinase") + 
        scale_color_gradient(low="gray", high="red")

p2 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_8486))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenosine Phosphorylase") + 
        scale_color_gradient(low="gray", high="blue")

p3 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4012))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenosine Kinase") + 
        scale_color_gradient(low="gray", high="blue")

p4 = ggplot(merged_umap, aes(x=UMAP_1, y=UMAP_2, color=HMR_4085))+ 
        geom_point(size=1, alpha=0.5) +
        theme_minimal() +
        labs(title="Adenine Phosphoribosyltransferase") + 
        scale_color_gradient(low="gray", high="blue")
```

```{r, fig.width=12, fig.height=6}
scater::multiplot(q1, p2, q2, p3, q3, p4, cols=3)
```















Let's create a seurat object
```{r}
rownames(flux_data) = flux_data$ID
flux_data$ID = NULL
neutrophil = CreateSeuratObject(counts=flux_data, 
                                project="neutrophil_flux", 
                                min.cells=1, 
                                min.features=20)
```



# 3. Data Preprocessing
Now that we hacked the Seurat object, let's run the usual preprocessing pipeline. Note that SCTransform doesn't work with this data.

```{r}
neutrophil = FindVariableFeatures(object=neutrophil, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)

# Log normalize data
neutrophil = NormalizeData(neutrophil, 
                     normalization.method='LogNormalize', 
                     scale.factor=10000)

all.rxns = rownames(neutrophil)
neutrophil = ScaleData(neutrophil, 
                      features=all.rxns)

# Run PCA
neutrophil = RunPCA(object=neutrophil,
              npcs=50,
              verbose=FALSE)

# Find shortest nearest neighbors and clusters
neutrophil = FindNeighbors(neutrophil, 
                     dims=1:50,
                     reduction='pca')

neutrophil = FindClusters(neutrophil, 
                    resolution=0.5)

# Perform UMAP
neutrophil = RunUMAP(neutrophil, 
               dims=1:50)
```

Create the DimPlot.
```{r,fig.height=12}
# Plot based on predicted clusters
DimPlot(neutrophil, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Neutrophil metabolic flux UMAP") +
        theme_minimal() +
        theme(plot.title = element_text(size=25),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

Get all metabolic reactions by predicted UMAP clusters
```{r}
cluster_markers = FindAllMarkers(neutrophil,
                         only.pos=TRUE,
                         min.pct=0.25,
                         logfc.threshold=0.25,
                         test.use="MAST",
                         )
cluster_markers = cluster_markers[!duplicated(cluster_markers[, "gene"]), ]

#writexl::write_xlsx(cluster_markers, "D:/Analysis/Immune/neutrophil_flux_clusters.xlsx")
```

Change the identity by day
```{r}
# Create color scheme by changing Identifier
neutrophil.day = SetIdent(obj=neutrophil, 
                          value=labels)

#tmp = AverageExpression(neutrophil.day, features=day_markers$gene)
levels(neutrophil.day) = c("Neutrophils  0", "Neutrophils  7", "Neutrophils  14", "Neutrophils  21")
DoHeatmap(object=neutrophil.day,
          features=day_markers$gene,
          group.colors=c('#E1E5F2', '#BFDBF7', '#1F7A8C', '#022B3A')) +
        scale_y_discrete(labels=day_markers$Name) +
        scale_fill_gradientn(colors=c("blue", "gray", "red"))
        
```

# 4. Run UMAP with the flux data
Rows correspond to different cells in each time point. Columns correspond to different metabolic reactions.
```{r,fig.height=4}
day_markers = FindAllMarkers(neutrophil.day,
                         only.pos=TRUE,
                         min.pct=0.25,
                         logfc.threshold=0.25,
                         test.use="MAST",
                         )

day_markers$Name = c("Adenosine Phosphorylase", 
                         "Adenine Phosphoribosyltransferase",
                         "Adenosine Kinase",
                         "Glutamate Dehydrogenase",
                         "Isocitrate:NAD+ Oxidoreductase",
                         "Isocitrate:NADP+ Oxidoreductase",
                         "Hypoxanthine Phosphoribosyltransferase",
                         "Inosine Phosphorylase",
                         "Inosine Kinase"
                         )

neutrophil_d0 = day_markers[day_markers$cluster=="Neutrophils  0",]$gene
FeaturePlot(neutrophil.day, 
            features=neutrophil_d0[1],
            label=FALSE,
            repel=TRUE,
            label.size=4) + 
        labs(title="Day 0", x="UMAP 1", y="UMAP 2")

neutrophil_d21 = day_markers[day_markers$cluster=="Neutrophils  21",]$gene
FeaturePlot(neutrophil.day, 
            features=neutrophil_d21[1],
            label=FALSE,
            repel=TRUE,
            label.size=4) + 
        labs(title="Day 21", x="UMAP 1", y="UMAP 2")
```
