---
title: "Immune cell differential metabolic gene expression analysis"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
This notebook performs differential gene expression analysis across all cell types. Additionally, it extracts metabolic genes corresponding to a metabolic reconstruction.

# 1. Load libraries
```{r}
library(mgcv)
library(splines)
library(Seurat)
library(tidyverse)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(phateR)
library(reticulate)
#install.packages('readxl')
library(readxl)
```

# 2. Separate out cell types
The following code creates lists of different cell types.
```{r}
celltype.list = SplitObject(lung, split.by="CellType")
celltypes = c("Stroma", "Alveolar Macrophages", "Macrophages", "Monocytes",
              "Pneumocytes", "B cells", "Granulocytes", "T cells",
              "Dendritic Cells", "Neutrophils", "NK cells", "Endothelial",
              "HSCs")
```

# 3. Perform Differential expression analysis
For this analysis, I compared each time point sequentially. From subpopulation analysis, I can do a more fine-course DEG, depending on cell types. This will be done in a later step.

**NOTE:** This code will take awhile to run for each time point and cell type. Get some coffee here.
```{r}
# Differential expression for different time points across different cell types
t0007_de = list()
t0714_de = list()
t1421_de = list()
for(i in 1:length(celltype.list)){
  tissuedata = SetIdent(celltype.list[[i]], value="Time")
  t0007_de[[i]] = FindMarkers(tissuedata, 
                              ident.1=0, indent.2=7,
                              min.cells.feature = 2, min.cells.group = 2)
  
  t0714_de[[i]] = FindMarkers(tissuedata, 
                              ident.1=7, indent.2=14,
                              min.cells.feature = 2, min.cells.group = 2)
  t1421_de[[i]] = FindMarkers(tissuedata, 
                              ident.1=14, indent.2=21,
                              min.cells.feature = 2, min.cells.group = 2)
}
```

# 4. Map mouse genes to human ortholog gene symbols

First, let's create a function `convertMouseGeneList`, which takes a dataframe and maps mouse gene symbols to human orthologs.

**NOTE:** This code may also take awhile. Time to go for a walk.
```{r}
# Install necessary libraries
install.packages("BiocManager")
BiocManager::install("biomaRt")
library(biomaRt)

# Create the mouse and human database
human = useMart("ensembl", dataset="hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset="mmusculus_gene_ensembl")

# Make a function that maps genes using two reference databases
convertMouseGeneList = function(gene_obj, db1, db2){
  orthologs = getLDS(attributes=c("mgi_symbol"), filters="mgi_symbol", 
                   values=rownames(gene_obj), mart=db2, 
                   attributesL=c("hgnc_symbol", "entrezgene_id"), martL=db1, 
                   uniqueRows=TRUE)
  mapped_gene_obj = merge(orthologs, gene_obj, 
                          by.x='MGI.symbol', by.y='row.names')
  return(mapped_gene_obj)
}
```

This code block takes those lists and maps the mouse to human orthologs using biomaRt.
```{r}
for(i in 1:length(celltype.list)){
  t0007_de[[i]] = convertMouseGeneList(t0007_de[[i]], human, mouse)
  t0714_de[[i]] = convertMouseGeneList(t0714_de[[i]], human, mouse)
  t1421_de[[i]] = convertMouseGeneList(t1421_de[[i]], human, mouse)
}
```

# 5. Get metabolic genes from metabolic reconstruction

## A) RECON1
First, let's read in metabolic genes from RECON1. We'll be using those 
```{r}
metabolic_map = read.csv("C:/Users/scott/Data/Mappings/GeneIDs/GeneSymbol_to_entrez.csv")
metabolic_map = unique(metabolic_map[, 1:2])
```

This function maps metabolic genes from the scRNASeq dataset. 
```{r}
getMetabolicGenes = function(df, map){
  mapped_df = merge(df, map, 
                    by.x='NCBI.gene..formerly.Entrezgene..ID', by.y='X_id')
  return(mapped_df)
}
```

This code block gets metabolic genes that intersect with the RECON1 reconstruction.
```{r}
mapped_0007 = list()
mapped_0714 = list()
mapped_1421 = list()
for(i in 1:length(celltype.list)){
  mapped_0007[[i]] = getMetabolicGenes(t0007_de[[i]], metabolic_map)
  mapped_0714[[i]] = getMetabolicGenes(t0714_de[[i]], metabolic_map)
  mapped_1421[[i]] = getMetabolicGenes(t1421_de[[i]], metabolic_map)
}
```

## B) iHUMAN
In this analysis, we'll use the iHUMAN metabolic reconstruction instead of RECON1. This reconstruction is much more comprehensive, and may reveal more interesting metabolic differences. Let's read in the metabolic gene map:
```{r}
# ACLX -> NEED TO CHANGE PATH
#metabolic_map = read.csv("C:/Users/scott/Data/Mappings/GeneIDs/GeneSymbol_to_entrez.csv")

# Linux
metabolic_map = read_excel("~/Data/Reconstructions/Human1/Human1_SynGO.xlsx")

# Get unique identifiers using the metabolic map
metabolic_map = unique(metabolic_map)
```

This function maps metabolic genes from the scRNASeq dataset. 
```{r}
getMetabolicGenes = function(df, map){
  mapped_df = merge(df, map, 
                    by.x='NCBI.gene..formerly.Entrezgene..ID', by.y='entrezgene')
  return(mapped_df)
}
```

Then let's get the mapped genes:
```{r}
mapped_0007 = list()
mapped_0714 = list()
mapped_1421 = list()
for(i in 1:length(celltype.list)){
  mapped_0007[[i]] = getMetabolicGenes(t0007_de[[i]], metabolic_map)
  mapped_0714[[i]] = getMetabolicGenes(t0714_de[[i]], metabolic_map)
  mapped_1421[[i]] = getMetabolicGenes(t1421_de[[i]], metabolic_map)
}
```
# 6. Get summary of differentially expressed metabolic genes
Now it would be interesting to get a high level overview of the number of up- and down-regulated metabolic genes from this analysis for each time point. 

Let's set the significance threshold to be 0.05.

First, let's get all significant metabolic genes. the data is still nested into a list.
```{r}
sig_0007 = list()
sig_0714 = list()
sig_1421 = list()
for(i in 1:length(celltype.list)){
  tmp1 = mapped_0007[[i]]
  tmp2 = mapped_0714[[i]]
  tmp3 = mapped_1421[[i]]
  
  sig_0007[[i]] = tmp1[tmp1[, 'p_val_adj'] < 0.05, ]
  sig_0714[[i]] = tmp2[tmp2[, 'p_val_adj'] < 0.05, ]
  sig_1421[[i]] = tmp3[tmp3[, 'p_val_adj'] < 0.05, ]
}
```

Now, we'll un-nest the list by concatenating these dataframes.
```{r}
df_0007 = list()
df_0714 = list()
df_1421 = list()

for(i in 1:length(celltype.list)){
  tmp1 = sig_0007[[i]]
  tmp2 = sig_0714[[i]]
  tmp3 = sig_1421[[i]]
  
  name1 = rep(celltypes[i], nrow(tmp1))
  name2 = rep(celltypes[i], nrow(tmp2))
  name3 = rep(celltypes[i], nrow(tmp3))
  
  tmp1[, "Tisssue"] = name1
  tmp2[, "Tisssue"] = name2
  tmp3[, "Tisssue"] = name3
  
  df_0007[[i]] = tmp1
  df_0714[[i]] = tmp2
  df_1421[[i]] = tmp3
  
}

final_df_0007 = dplyr::bind_rows(df_0007)
final_df_0714 = dplyr::bind_rows(df_0714)
final_df_1421 = dplyr::bind_rows(df_1421)
```

Let's vertically concatenate the dataframes:
```{r}
final_df_0007$Time = "Day7/Day0"
final_df_0714$Time = "Day14/Day7"
final_df_1421$Time = "Day21/Day14"

tmp = rbind(final_df_0007, final_df_0714)
final_df = rbind(tmp, final_df_1421)

# Linux
write.csv(x=final_df, file="~/Analysis/Immune/iHUMAN_DEG_Aggregate.csv", row.names=TRUE)
```

# 7. Individual Cell Type Metabolic Gene Expression Analysis
This analysis is slightly different, in that we'll be doing the following to **get differentially expressed metabolic genes for individual cells.**

## Without Data Imputation
First, let's try this kind of analysis without data imputation methods. We'll do that by computing gene-wise Z-scores and computing a p-value associated with the Z-scores from a Normal distribution.

Specifically, we'll perform the following steps:

  1. Grab count matrices corresponding to individual cell types.
  2. Compute gene-wise Z-scores for each cell type and each time point.
  3. Compute corresponding P-values for each data matrix.
  4. Perform BH correction.
  5. Resize the matrix
  6. Re-map to metabolic genes

```{r}
install.packages('phateR')
library(phateR)
library(Matrix)
reticulate::conda_remove("r-reticulate")
reticulate::py_discover_config(required_module = "phate")

for(i in 1:length(celltype.list)){
  tmp = as.matrix(celltype.list[[i]]@assays$RNA@counts)
  tmp = phateR::phate(tmp)
  tmp = apply(tmp, 1, scale)
  
  # Map to metabolic genes
  mapped_0007[[i]] = getMetabolicGenes(t0007_de[[i]], metabolic_map)
  mapped_0714[[i]] = getMetabolicGenes(t0714_de[[i]], metabolic_map)
  mapped_1421[[i]] = getMetabolicGenes(t1421_de[[i]], metabolic_map)
}
```

## With Data Imputation

  1. Use MAGIC to perform data imputation for dropped out cells
  2. Compute Z-scores and corresponding p-values for each gene across each cell.
  3. Perform BH-adjustment for p-values
  4. Save data matrix for COBRA.