---
title: "Immune cell pseudotime analysis"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
This notebook performs differential gene expression analysis across all cell types. Additionally, it extracts metabolic genes corresponding to a metabolic reconstruction.

# 1. Load libraries
```{r}
#install.packages(c("mgcv", "splines", "Seurat", "tidyverse", 
#                   "viridis", "matrixStats", "UpSetR", 
#                   "reshape2", "ggplot2", "phateR"), dependencies=TRUE)

library(mgcv)
library(splines)
library(Seurat)
library(tidyverse)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(phateR)
library(reticulate)
#install.packages('readxl')
library(readxl)

# Monocle3 is still in beta
#BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
#                       'limma', 'S4Vectors', 'SingleCellExperiment',
#                       'SummarizedExperiment', 'batchelor', 'Matrix.utils'))
#install.packages("devtools")
library(devtools)
#devtools::install_github('cole-trapnell-lab/leidenbase', dependencies=TRUE)
#devtools::install_github('cole-trapnell-lab/monocle3', dependencies=TRUE)

#remotes::install_github('satijalab/seurat-wrappers')

library(SeuratWrappers)
library(monocle3)
library(Matrix)
```

# 2. Set up data objects for Monocle

We need to transform the Seurat object to a Cell Data Set, that can be used for Monocle and other packages.
```{r}
rm(list=ls())

# ACLX
#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# Dell
load("D:/Data/scRNASeq/shea/lung.robj")

# Linux
#load("~/Data/scRNASeq/shea/lung.Robj")
```

Let's load it up as a cds.
```{r}
dat = as.matrix(lung@assays$RNA@data)
cell_metadata = data.frame(lung@active.ident)
lung_cds = new_cell_data_set(dat, 
                             cell_metadata=cell_metadata)
#lung_cds$cell.type = cell_metadata
```
# 3. Correct for batch effects
First, we'll preprocess the lung data using the `preprocess_cds` and `reduce_dimension` methods.
```{r}
lung_cds = preprocess_cds(lung_cds, num_dim=50)
lung_cds = align_cds(lung_cds, alignment_group="batch")
```

# 4. Dimension reduction
We'll run PCA and UMAP.
```{r}
lung_cds = reduce_dimension(lung_cds, reduction_method='PCA')
lung_cds = reduce_dimension(lung_cds, reduction_method='UMAP')
```

# 5. Perform Clustering 
Let's perform some clustering and visualize the clusters.
```{r}
lung_cds = cluster_cells(lung_cds, resolution=1e-5)
plot_cells(lung_cds, color_cells_by="partition")
```

Let's learn the trajectory algorithmically.
```{r}
lung_cds = learn_graph(lung_cds)
plot_cells(lung_cds,
           color_cells_by = "partition",
           label_groups_by_cluster=TRUE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
```

# 6. Construct the Pseudotime Trajectories
This is a function to identify the root principle points automatically.
```{r}
# a helper function to identify the root principal points:
get_earliest_principal_node = function(cds, time_bin="130-170"){
  cell_ids = which(colData(cds)[, "embryo.time.bin"] == time_bin)
  
  closest_vertex = cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex = as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes =
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numerics(names(which.max(table(closest_vertex[cell_ids,]))))]
  return(root_pr_nodes)
}
```


```{r}
ldha_cds = order_cells(ldha_cds, root_pr_nodes=get_earliest_principal_node(ldha_cds))
```

```{r}
ldha_cds = order_cells(ldha_cds)
plot_genes_in_pseudotime(ldha_cds,
                         min_expr=0.5)
```

