---
title: "Immune cell subpopulation using metabolism data"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
**This analysis contains 3 time points:**

  1. *Day 7*:  Pre-metastatic niche - tumor cells have not arrived
  2. *Day 14*: Metastatic niche - tumor cells arrived but no proliferation
  3. *Day 21*: Metabstatic niche - secondary tumors in lung 

**The driving questions are the following:**

  1. Do neutrophils adopt an immunosuppressive or immunoevasive phenotype driven by rewired metabolism?
  2. How do other immune cell types, such as NK cell and T-cell, change their metabolic pathways to facilitate immunosuppression and evasion?
  3. Can we infer a regulatory relationship between neutrophil and T-cells via altered metabolic processes?

# 1. Load libraries and data
First, we'll load the libraries and dataset we'll need to analyze the data.

```{r}
# Load the libraries
#install.packages(c("mgcv", "splines", "Seurat", "tidyverse", 
#                   "viridis", "matrixStats", "UpSetR", 
#                   "reshape2", "ggplot2", "phateR"))
library(mgcv)
library(splines)
library(Seurat)
library(tidyverse)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(phateR)
library(reticulate)

```

```{r}
rm(list=ls())
# Load the data
# Dell
#load("D:/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# ACLX
load("C:/Users/scott/Data/scRNASeq/shea/lung_human.robj", verbose=TRUE)
#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# Linux
#load("~/Data/scRNASeq/shea/lung.Robj")
#unique(neutrophils@meta.data[["CellType"]])
```

# 2. Split data to be separate cell lines
We'll first split the data to be separate lineages. I will analyze 4 specifically:
  1. Neutrophils
  2. NK cells
  3. T cells
  4. Granulocytes
  
```{r}
#orig.indent = lung_metabolism@active.ident
#lung_metabolism@meta.data$CellType = orig.indent
lung_metabolism@active.ident = lung_metabolism@meta.data$CellType
Idents(lung_metabolism) = lung_metabolism@meta.data$CellType
celltype.list = SplitObject(lung_metabolism, split.by="ident")

# Create separate Seurat objects to visualize.
neutrophils   = celltype.list$Neutrophils
nk            = celltype.list$`NK Cells`
tcell         = celltype.list$`T Cells`
gran          = celltype.list$Granulocytes
```

# 3. Create function that will perform all downstream preprocessing steps.
For more in-depth sanity checking, you can check out the `eda_all_genes.Rmd` file. In short, the following is performed:
  1. Identify variable features
  2. Log normalize the data
  3. Scale the data 
  4. Run PCA
  5. Identify shortest nearest neighbors and clusters
  6. Perform UMAP
  
```{r}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  # Log normalize data
  gene_obj = NormalizeData(gene_obj, 
                       normalization.method='LogNormalize', 
                       scale.factor=10000)
  
  # Scale data
  all.genes = rownames(gene_obj)
  gene_obj = ScaleData(gene_obj, 
                   features=all.genes)
  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                npcs=50,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:50,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:50)
  
  return(gene_obj)
}
```

# 4. Perform subsetting on Neutrophils
This analysis subsets based on predicted clusters.
```{r, fig.width = 12, fig.height = 12}
neutrophils = find_subpopulations(neutrophils)

# Plot based on predicted clusters
DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Neutrophil UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(neutrophils,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(neutrophils, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(neutrophils, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

Let's also visualize based on time.
```{r, fig.width = 12, fig.height = 12}
# Create color scheme by changing Identifier
expt = neutrophils@meta.data["Experiment"]
expt = as.character(expt$Experiment)
neutrophils = SetIdent(obj=neutrophils, 
                       value=expt)

# Plot based on time-course
DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Neutrophil UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(neutrophils,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(neutrophils, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(neutrophils, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

# 5. Perform subsetting on NK Cells
This analysis subsets based on predicted clusters and time.
```{r, fig.width = 12, fig.height = 12}
nk = find_subpopulations(nk)

# Plot based on predicted clusters
DimPlot(nk, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="NK-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(nk,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(nk, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(nk, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

```{r, fig.width = 12, fig.height = 12}
# Create color scheme by changing Identifier
expt = nk@meta.data["Experiment"]
expt = as.character(expt$Experiment)
nk = SetIdent(obj=nk, value=expt)

# Plot based on time-course
DimPlot(nk, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="NK-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(nk,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(nk, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(nk, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

# 6. Perform subsetting on T Cells
This analysis subsets based on predicted clusters and time.
```{r, fig.width = 12, fig.height = 12}
tcell = find_subpopulations(tcell)

# Plot based on predicted clusters
DimPlot(tcell, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="T-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(tcell,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(tcell, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(tcell, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

```{r, fig.width = 12, fig.height = 12}
# Create color scheme by changing Identifier
expt = tcell@meta.data["Experiment"]
expt = as.character(expt$Experiment)
tcell = SetIdent(obj=tcell, value=expt)

# Plot based on time-course
DimPlot(tcell, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="T-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(tcell,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(tcell, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(tcell, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

# 7. Perform subsetting on Granulocytes
```{r, fig.width = 12, fig.height = 12}
gran = find_subpopulations(gran)

# Plot based on predicted clusters
DimPlot(gran, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=2) + 
        labs(x="UMAP 1", y="UMAP 2", title="Granulocytes UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}
markers = FindAllMarkers(gran,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(gran, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(gran, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

```{r, fig.width = 12, fig.height = 12}
# Create color scheme by changing Identifier
expt = gran@meta.data["Experiment"]
expt = as.character(expt$Experiment)
gran = SetIdent(obj=gran, value=expt)

# Plot based on time-course
DimPlot(gran, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Granulocyte UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r, fig.width = 12, fig.height = 12}

markers = FindAllMarkers(gran,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(gran, 
          features=markers$gene,
          label=FALSE)

top5 = markers %>% 
          group_by(cluster) %>% 
          top_n(n=5, 
                wt=avg_logFC)

# Feature plots
FeaturePlot(gran, 
            features=top5$gene,
            label=TRUE,
            repel=TRUE,
            label.size=4,
            ncol=5)
```

# 8. Conclusions

**From this analysis, I found that:**

  1. Neutrophils have two distinct populations that are dependent on time. Thus, this corresponds to benign / metastatic phenotypes. 
  2. There is no distinguishable difference between NK cells as a function of time.
  3. There are 3 different T-cell populations that do not differ as a function of time.
  4. There are some granulocytes that do differ as a function of time, but as as a function of tissue lineage.
  
**To improve upon this analysis, I can do the following:**

  1. Hyperparameter tuning
  2. Different UMAP variants
  3. Run PHATE (attempted on R / can try on Python)
  4. Look at 3 dimensions rather than 2.