---
title: "Immune cell dimension reduction"
author: "Scott Campit"
output: 
  html_notebook:
    toc: yes
    theme: united
---

# Summary
**This analysis contains 3 time points:**

  1. *Day 7*:  Pre-metastatic niche - tumor cells have not arrived
  2. *Day 14*: Metastatic niche - tumor cells arrived but no proliferation
  3. *Day 21*: Metabstatic niche - secondary tumors in lung 

**The driving questions are the following:**

  1. Do neutrophils adopt an immunosuppressive or immunoevasive phenotype driven by rewired metabolism?
  2. How do other immune cell types, such as NK cell and T-cell, change their metabolic pathways to facilitate immunosuppression and evasion?
  3. Can we infer a regulatory relationship between neutrophil and T-cells via altered metabolic processes?

# 1. Load libraries and data
First, we'll load the libraries and dataset we'll need to analyze the data.

```{r}
# Load the libraries
#install.packages(c("mgcv", "splines", "Seurat", "tidyverse", 
#                   "viridis", "matrixStats", "UpSetR", 
#                   "reshape2", "ggplot2", "phateR"))
library(mgcv)
library(splines)
library(Seurat)
library(tidyverse)
library(viridis)
library(matrixStats)
library(UpSetR)
library(reshape2)
library(ggplot2)
library(phateR)
library(reticulate)
library(MAST)

```

```{r}
rm(list=ls())
# Load the data
# Dell
load("D:/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# ACLX
#load("C:/Users/scott/Data/scRNASeq/shea/lung.robj", verbose=TRUE)

# Linux
#load("~/Data/scRNASeq/shea/lung.Robj")
#unique(neutrophils@meta.data[["CellType"]])
```

# 2. Split data to be separate cell lines
We'll first split the data to be separate lineages. I will analyze 4 specifically:
  1. Neutrophils
  2. NK cells
  3. T cells
  4. Granulocytes
  
```{r}
orig.indent = lung@active.ident
celltype.list = SplitObject(lung, split.by="CellType")

# Create separate Seurat objects to visualize.
neutrophils   = celltype.list$Neutrophils
nk            = celltype.list$`NK Cells`
tcell         = celltype.list$`T Cells`
gran          = celltype.list$Granulocytes
```

# 3. Create function that will perform all downstream preprocessing steps.
For more in-depth sanity checking, you can check out the `eda_all_genes.Rmd` file. In short, the following is performed:
  1. Identify variable features
  2. Log normalize the data
  3. Scale the data 
  4. Run PCA
  5. Identify shortest nearest neighbors and clusters
  6. Perform UMAP
  
```{r}
find_subpopulations = function(gene_obj){
  
  # Identify variable features
  gene_obj = FindVariableFeatures(object=gene_obj, 
                              mean.function=ExpMean,
                              dispersion.function=LogVMR,
                              selection.method='vst',
                              x.low.cutoff=-Inf,
                              x.high.cutoff=Inf,
                              y.cutoff=Inf,
                              nfeatures = 2000)
  
  # Log normalize data
  gene_obj = NormalizeData(gene_obj, 
                       normalization.method='LogNormalize', 
                       scale.factor=10000)
  
  # Scale data
  all.genes = rownames(gene_obj)
  gene_obj = ScaleData(gene_obj, 
                   features=all.genes)
  
  gene_obj = SCTransform(gene_obj)

  
  # Run PCA
  gene_obj = RunPCA(object=gene_obj,
                npcs=50,
                verbose=FALSE)
  
  # Find shortest nearest neighbors and clusters
  gene_obj = FindNeighbors(gene_obj, 
                       dims=1:50,
                       reduction='pca')
  gene_obj = FindClusters(gene_obj, 
                      resolution=0.5)
  
  # Perform UMAP
  gene_obj = RunUMAP(gene_obj, 
                 dims=1:50)
  
  return(gene_obj)
}
```

# 4. Perform subsetting on Neutrophils
This analysis subsets based on predicted clusters.
```{r}
neutrophils = find_subpopulations(neutrophils)

# Plot based on predicted clusters
DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Neutrophil UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=25),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(neutrophils,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25,
                         test.use='MAST')
top5 = markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(neutrophils, 
          features=top5$gene)
```


## All clusters
```{r}
FeaturePlot(neutrophils, 
            features=c("Saa3", "B2m", "Fpr2", "Ltf", "Hbb-bt"),
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

## Cluster 0
These thresholds were selected arbitrarily.

Saa3: Up-regulation is associated with the suppression of neutrophil inflammatory response for cluster 0. 

  * Murdoch, C. C., Espenschied, S. T., Matty, M. A., Mueller, O., Tobin, D. M., & Rawls, J. F. (2019). Intestinal Serum amyloid A suppresses systemic neutrophil activation and bactericidal activity in response to microbiota colonization. PLoS pathogens, 15(3), e1007381.
  
Il1r2: Promotes self-renewal of reast tumor initiating cells and breast cancer proliferation/invasion

  * Zhang, L., Qiang, J., Yang, X., Wang, D., Rehman, A. U., He, X., ... & Liu, S. (2020). IL1R2 Blockade Suppresses Breast Tumorigenesis and Progression by Impairing USP15‐Dependent BMI1 Stability. Advanced Science, 7(1), 1901728.

```{r}
neutrophil_c0 = markers[markers$cluster==0 & markers$pct.1 > 0.55 & markers$pct.2 < 0.45,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c0,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

## Cluster 1

B2m: Marker for neutrophil 

  * Zhang, X., Ding, L., & Sandford, A. J. (2005). Selection of reference genes for gene expression studies in human neutrophils by real-time PCR. BMC molecular biology, 6(1), 1-7.

Marcks: In unstimulated neutrophils

  * Eckert, R. E., Neuder, L. E., Park, J., Adler, K. B., & Jones, S. L. (2010). Myristoylated alanine-rich C-kinase substrate (MARCKS) protein regulation of human neutrophil migration. American journal of respiratory cell and molecular biology, 42(5), 586-594.

Rgs2: Bachoprotective in acute neutrophilic inflammation and dampens Th1 responses

  * George, T., Chakraborty, M., Giembycz, M. A., & Newton, R. (2018). A bronchoprotective role for Rgs2 in a murine model of lipopolysaccharide-induced airways inflammation. Allergy, Asthma & Clinical Immunology, 14(1), 1-14.

Lyz2: Neutrophil marker

  * Kim, M. H., Yang, D., Kim, M., Kim, S. Y., Kim, D., & Kang, S. J. (2017). A late-lineage murine neutrophil precursor population exhibits dynamic changes during demand-adapted granulopoiesis. Scientific reports, 7(1), 1-15.

Plek: NaN


```{r, fig.height=12}
neutrophil_c1 = markers[markers$cluster==1 & markers$pct.1 > 0.55 & markers$pct.2 < 0.45,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c1,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

## Cluster 2

Stfa3: NaN

Ifitm6: Increased in macrophages of tumor bearing mice.

  * Han JH, Lee S, Park YS, et al. IFITM6 expression is increased in
macrophages of tumor-bearing mice. Oncol Rep 2011;25:531–536.

Fpr2: Promotes malignancy of human colon cancer.

  * Liang, W., Chen, K., Gong, W., Yoshimura, T., Le, Y., Wang, Y., & Wang, J. M. (2020). The contribution of chemoattractant GPCRs, formylpeptide receptors, to inflammation and cancer. Frontiers in endocrinology, 11, 17.
  
  * Chen, K., Bao, Z., Gong, W., Tang, P., Yoshimura, T., & Wang, J. M. (2017). Regulation of inflammation by members of the formyl-peptide receptor family. Journal of autoimmunity, 85, 64-77.

```{r}
neutrophil_c2 = markers[markers$cluster==2 & markers$pct.1 > 0.65 & markers$pct.2 < 0.35,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c2,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

## Cluster 3

Ltf: Lactotransferrin - found in secondary granules of neutrophils

  * Kruzel, M. L., Zimecki, M., & Actor, J. K. (2017). Lactoferrin in a context of inflammation-induced pathology. Frontiers in immunology, 8, 1438.
  
Camp: Promotes recriotment of inflammatory monocytes (atherosclerosis)

  * Döring, Y., Drechsler, M., Wantha, S., Kemmerich, K., Lievens, D., Vijayan, S., ... & Soehnlein, O. (2012). Lack of neutrophil-derived CRAMP reduces atherosclerosis in mice. Circulation research, 110(8), 1052-1056.
  
Cybb(NOX2): Overexpression leads to accrued ROS production leading to survival of cancer cells.

  * Weyemi, U., E Redon, C., R Parekh, P., Dupuy, C., & M Bonner, W. (2013). NADPH Oxidases NOXs and DUOXs as putative targets for cancer therapy. Anti-Cancer Agents in Medicinal Chemistry (Formerly Current Medicinal Chemistry-Anti-Cancer Agents), 13(3), 502-514.
  
  * Aydin, E., Johansson, J., Nazir, F. H., Hellstrand, K., & Martner, A. (2017). Role of NOX2-derived reactive oxygen species in NK cell–mediated control of murine melanoma metastasis. Cancer immunology research, 5(9), 804-811.
  
Aldh2: NaN

Cd177: Highly expressed in CRC and UC-associated cancers

  * Zhou, G., Peng, K., Song, Y., Yang, W., Shu, W., Yu, T., ... & Liu, Z. (2018). CD177+ neutrophils suppress epithelial cell tumourigenesis in colitis-associated cancer and predict good prognosis in colorectal cancer. Carcinogenesis, 39(2), 272-282.
  
  * Schiffmann, L. M., Fritsch, M., Gebauer, F., Günther, S. D., Stair, N. R., Seeger, J. M., ... & Coutelle, O. (2019). Tumour-infiltrating neutrophils counteract anti-VEGF therapy in metastatic colorectal cancer. British journal of cancer, 120(1), 69-78.
  
Ckap4: NaN

```{r,fig.height=12}
neutrophil_c3 = markers[markers$cluster==3 & markers$pct.1 > 0.75 & markers$pct.2 < 0.25,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c3,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

## Cluster 4
Most likely residual erythrocytes classified as neutrophils due to hemoglobin biomarkers.
```{r}
neutrophil_c4 = markers[markers$cluster==4 & markers$pct.1 > 0.65 & markers$pct.2 < 0.35,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c4,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

Let's also visualize based on time.
```{r}
# Create color scheme by changing Identifier
expt = neutrophils@meta.data["Experiment"]
expt = as.character(expt$Experiment)
neutrophils = SetIdent(obj=neutrophils, 
                       value=expt)
```

```{r}
# Plot based on time-course
DimPlot(neutrophils, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Neutrophil UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=25),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(neutrophils,
                         only.pos=TRUE,
                         min.pct=0.25,
                         logfc.threshold=0.25)

top5 = markers %>% 
  group_by(cluster) %>% 
  top_n(n=20, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(neutrophils, 
          features=top5$gene)

```

# Perform subpopulation analysis on metabolic subset

## Create function to get metabolic subset
This function grabs metabolic genes using Entrez IDs
```{r}
getMetabolicGenes = function(df, map){
  mapped_df = merge(df, map, 
                    by.x='NCBI.gene..formerly.Entrezgene..ID', by.y='entrezgene')
  return(mapped_df)
}
```

## Create library to map human and mouse genes
Now, let's get the genes, map it to human identifiers, and map that to iHUMAN.
```{r}
# Install necessary libraries
#install.packages("BiocManager")
#BiocManager::install("biomaRt")
library(biomaRt)

# Create the mouse and human database
human = useMart("ensembl", dataset="hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset="mmusculus_gene_ensembl")
```
## Map human and mouse genes
Create a function that maps MGI gene symbols to HGNC gene symbols and Entrez IDs.
```{r}
# Make a function that maps genes using two reference databases
convertMouseGeneList = function(gene_obj, db1, db2){
  orthologs = getLDS(attributes=c("mgi_symbol"), filters="mgi_symbol", 
                   values=rownames(gene_obj), mart=db2, 
                   attributesL=c("hgnc_symbol", "entrezgene_id"), martL=db1, 
                   uniqueRows=TRUE)
  mapped_gene_obj = merge(orthologs, gene_obj, 
                          by.x='MGI.symbol', by.y='row.names')
  return(mapped_gene_obj)
}
```

## Grab metabolic genes that are signficant and map to human genes.
Now let's get our count matrix and perform this mapping.
```{r}
# Convert mouse to human genes
neutrophil_metab_mat = convertMouseGeneList(markers, human, mouse)

# iHUMAN reconstruction
human_map = readxl::read_excel("D:/Data/Reconstructions/Human1/Human1_SynGO.xlsx")

# Grab metabolic genes
neutrophil_metab_mat = getMetabolicGenes(neutrophil_metab_mat, human_map)
```

# Analyze metabolic genes in clusters
## Cluster 0
```{r}
neutrophil_c0 = neutrophil_metab_mat[neutrophil_metab_mat$cluster==0 & markers$pct.1 > 0.30 & markers$pct.2 < 0.30,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c0,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```
## Cluster 1
```{r, fig.height=12}
neutrophil_c1 = neutrophil_metab_mat[neutrophil_metab_mat$cluster==1 & markers$pct.1 > 0.30 & markers$pct.2 < 0.30,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c1,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```
## Cluster 2
```{r, fig.height=12}
neutrophil_c2 = neutrophil_metab_mat[neutrophil_metab_mat$cluster==2 & markers$pct.1 > 0.30 & markers$pct.2 < 0.30,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c2,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```
## Cluster 3
```{r, fig.height=12}
neutrophil_c3 = neutrophil_metab_mat[neutrophil_metab_mat$cluster==3 & markers$pct.1 > 0.30 & markers$pct.2 < 0.30,]$gene
FeaturePlot(neutrophils, 
            features=neutrophil_c3,
            label=TRUE,
            repel=TRUE,
            label.size=4)
```

# 5. Perform subsetting on NK Cells
This analysis subsets based on predicted clusters and time.
```{r}
nk = find_subpopulations(nk)

# Plot based on predicted clusters
DimPlot(nk, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="NK-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(nk,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(nk, 
          features=markers$gene)
```

```{r}
# Create color scheme by changing Identifier
expt = nk@meta.data["Experiment"]
expt = as.character(expt$Experiment)
nk = SetIdent(obj=nk, value=expt)

# Plot based on time-course
DimPlot(nk, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="NK-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(nk,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(nk, 
          features=markers$gene)
```

# 6. Perform subsetting on T Cells
This analysis subsets based on predicted clusters and time.
```{r}
tcell = find_subpopulations(tcell)

# Plot based on predicted clusters
DimPlot(tcell, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="T-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(tcell,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(tcell, 
          features=markers$gene)
```

```{r}
# Create color scheme by changing Identifier
expt = tcell@meta.data["Experiment"]
expt = as.character(expt$Experiment)
tcell = SetIdent(obj=tcell, value=expt)

# Plot based on time-course
DimPlot(tcell, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="T-cell UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(tcell,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(tcell, 
          features=markers$gene)
```

# 7. Perform subsetting on Granulocytes
```{r}
gran = find_subpopulations(gran)

# Plot based on predicted clusters
DimPlot(gran, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        pt.size=2) + 
        labs(x="UMAP 1", y="UMAP 2", title="Granulocytes UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(gran,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(gran, 
          features=markers$gene)
```

```{r}
# Create color scheme by changing Identifier
expt = gran@meta.data["Experiment"]
expt = as.character(expt$Experiment)
gran = SetIdent(obj=gran, value=expt)

# Plot based on time-course
DimPlot(gran, 
        reduction="umap",
        dims=c(1, 2),
        label=TRUE,
        repel=TRUE,
        label.size=4,
        na.value = "grey50") + 
        labs(x="UMAP 1", y="UMAP 2", title="Granulocyte UMAP compression") +
        theme_minimal() +
        theme(plot.title = element_text(size=30),
              legend.text=element_text(size=20),
              axis.title.x=element_text(size=20),
              axis.title.y=element_text(size=20),
              axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20)) +
        geom_hline(yintercept=0) +
        geom_vline(xintercept=0)
```

We should also report the biomarkers.
```{r}
markers = FindAllMarkers(gran,
                         only.pos=TRUE,
                         min.pct = 0.25,
                         logfc.threshold=0.25)
markers %>% 
  group_by(cluster) %>% 
  top_n(n=5, 
        wt=avg_logFC)

# Create the heatmap of top markers
DoHeatmap(gran, 
          features=markers$gene)
```

# 8. Conclusions

**From this analysis, I found that:**

  1. Neutrophils have two distinct populations that are dependent on time. Thus, this corresponds to benign / metastatic phenotypes. 
  2. There is no distinguishable difference between NK cells as a function of time.
  3. There are 3 different T-cell populations that do not differ as a function of time.
  4. There are some granulocytes that do differ as a function of time, but as as a function of tissue lineage.
  
**To improve upon this analysis, I can do the following:**

  1. Hyperparameter tuning
  2. Different UMAP variants
  3. Run PHATE (attempted on R / can try on Python)
  4. Look at 3 dimensions rather than 2.